core/src/mender-client.c

```diff
@@ -1049,6 +1055,50 @@ mender_prepare_new_provides(mender_artifact_ctx_t *mender_artifact_ctx, char **n
 
     return MENDER_OK;
 }
+
+static mender_err_t
+mender_check_device_compatibility(mender_artifact_ctx_t *mender_artifact_ctx) {
+
+    /* We need to load the stored provides */
+    mender_key_value_list_t *stored_provides = NULL;
+    if (MENDER_FAIL == mender_storage_get_provides(&stored_provides)) {
+        return MENDER_FAIL;
+    }
+
+    mender_err_t ret = MENDER_FAIL;
+
+    /* Get depends */
+    mender_key_value_list_t *depends = NULL;
+    for (size_t i = 0; i < mender_artifact_ctx->payloads.size; i++) {
+        if (MENDER_OK != mender_utils_append_list(&depends, &mender_artifact_ctx->payloads.values[i].depends)) {
+            mender_log_error("Unable to append depends");
+            goto END;
+        }
+    }
+
+    /* Match depends from artifact with device's provides */
+    for (mender_key_value_list_t *depends_item = depends; NULL != depends_item; depends_item = depends_item->next) {
+        bool matches = false;
+        for (mender_key_value_list_t *provides_item = stored_provides; NULL != provides_item; provides_item = provides_item->next) {
+            /* Match key-value from depends with provides */
+            if (0 == strncmp(depends_item->key, provides_item->key, strlen(depends_item->key))
```

Comment:
Since these are null-byte terminated string, you may as well use `strcmp()`. Using `strlen()` and `strncmp()` causes the string to be iterated twice, instead of once.

Reply:
And I guess it's time to introduce `#define StringEqual(str1, str2)...`